example/gosumtype_1_declaration.boilerplate.go

```diff
@@ -126,3 +199,89 @@ func UserMap[A any](value User, variants UserVariantsMap[A]) A {
 	})
 	return result
 }
+
+// User = Anonymous | Member | Admin
+type User struct {
+	user user
+}
+
+// user is the interface for UserVariants
+type user interface {
+	Match(variants UserVariants)
+}
+
+func (s User) Match(variants UserVariants) {
+	s.user.Match(variants)
+}
+func (s User) MarshalJSON() (data []byte, err error) {
+	s.user.Match(UserVariants{
+		Anonymous: func(arg0Arg PaymentMethod) {
+			data, err = json.Marshal([]any{
+				"Anonymous",
+				anonymousUserVariants{
+					Arg0: arg0Arg,
+				}})
+		},
+		Member: func(emailArg string, sinceArg time.Time) {
+			data, err = json.Marshal([]any{
+				"Member",
+				memberUserVariants{
+					Email: emailArg,
+					Since: sinceArg,
+				}})
+		},
+		Admin: func(emailArg string) {
+			data, err = json.Marshal([]any{
+				"Admin",
+				adminUserVariants{
+					Email: emailArg,
+				}})
+		},
+	})
+	return data, err
+}
+func (s *User) UnmarshalJSON(data []byte) error {
+	// The expected format is ["TypeName", { ... data... }]
+	var raw []json.RawMessage
+	if err := json.Unmarshal(data, &raw); err != nil {
+		return fmt.Errorf("expected an array with type and data, got error: %w", err)
+	}
+	if len(raw) != 2 {
+		return fmt.Errorf("expected array of two elements [type, data], got %d elements", len(raw))
+	}
+	// Unmarshal the first element to get the type
+	var typeName string
+	if err := json.Unmarshal(raw[0], &typeName); err != nil {
+		return fmt.Errorf("failed to unmarshal type name: %w", err)
+	}
+	switch typeName {
+	case "Anonymous":
```

Comment:
json: when we decode, we first decode the first string to find our variant. then we decode the second item in the array accordingly